Refactor LoadMD data indexing and memory algorithms to speed up processing

williamfgc opened the pull request:
**Description of work.**

Changes targeting LoadMD algorithms efficiency.
Refactor class hierarchy to inherit from `Mantid::Kernel::NexusHDF5Descriptor` and resuse metadata indexing.
Refactor metadata index processing using NexusHDF5Descriptor to avoid regenerating std::map at every level.
Changes in memory allocation reusing containers as much as possible.
Reported speed ups on 20% on average on ORNL SNS/HFIR production system for 180 experiments, it also reduces system variability effects.

**To test:**
Refactoring should pass existing CI tests, including system tests.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

williamfgc committed:
Refactor LoadMD class hierarchy structure

Refactor confidence algorithm

williamfgc committed:
Trying out different files

williamfgc committed:
Add root dir in tests

To test on different platforms

williamfgc committed:
Reverting tests

Use allEntries to find MDEventWorkspace or MDHistoWorkspace

williamfgc committed:
Refactor first getEntries call in execLoader

williamfgc committed:
Overload MDBoxFlatTree::loadExperimentInfos

Version that passes a const
std::shared_ptr<Mantid::Kernel::NexusHDF5Descriptor>
Replaces getEntries metadata reconstruction

williamfgc committed:
Refactor structure up to Run::loadNexus

To do:
Refactor Run::loadNexus and LogManager::loadNexus

williamfgc committed:
Replace map entries in Run::loadNexus

Loop through all entries instead of forming std::map
Pass the correct prefix for current group

williamfgc committed:
Overload Run::loadNexus and LogManager::loadNexus

Introduce Mantid::Kernel::NexusHDF5Descriptor for quick look up

williamfgc committed:
LoadMDTest modified

williamfgc committed:
Modified LoadMDTest

williamfgc committed:
Make m_tableData a member to avoid reallocation

williamfgc committed:
Overload PropertyNexus::loadProperty

Replace last getEntries call in LoadMD
Switch LogManager::loadNexus to empty as it must be overloaded.

williamfgc committed:
Replace local vector with temporary

Reduce memory allocations

williamfgc committed:
Reuse tokens internal memory

Avoid reallocation when reading parameter maps

williamfgc committed:
Reuse memory in string tokenizer

Changes to all local scope functions to reuse member memory
Avoiding  frequent reallocation

williamfgc committed:
Move const variable outside loop

williamfgc committed:
Restore original LoadMDTest

williamfgc committed:
Address clang-format CI issue

Remove unused function

quantumsteve commented on Framework/API/inc/MantidAPI/ExperimentInfo.h:
```diff
@@ -120,6 +121,11 @@ class MANTID_API_DLL ExperimentInfo {
   void saveExperimentInfoNexus(::NeXus::File *file, bool saveInstrument, bool saveSample, bool saveLogs) const;
+
+  void loadExperimentInfoNexus(const std::string &nxFilename, ::NeXus::File *file, std::string &parameterStr,
+                               const std::shared_ptr<Mantid::Kernel::NexusHDF5Descriptor> &fileInfo,
```
Can we pass fileInfo as a reference (`Mantid::Kernel::NexusHDF5Descriptor &`) in this and similar functions?

williamfgc replied:
Good call.

quantumsteve commented on Framework/DataObjects/src/MDBoxFlatTree.cpp:
```diff
@@ -378,6 +379,90 @@ void MDBoxFlatTree::saveExperimentInfos(::NeXus::File *const file, const API::IM
   }
 }
 
+void MDBoxFlatTree::loadExperimentInfos(::NeXus::File *const file, const std::string &filename,
```
Can we factor our portions of this member function that are the same as in `void MDBoxFlatTree::loadExperimentInfos(::NeXus::File *const file, const std::string &filename,
                                        const std::shared_ptr<Mantid::API::MultipleExperimentInfos> &mei, bool lazy)`?

quantumsteve commented on Framework/Kernel/src/PropertyNexus.cpp:
```diff
@@ -108,6 +108,107 @@ std::unique_ptr<Property> makeStringProperty(::NeXus::File *file, const std::str
+ * @param prefix :: current absolute target prefix
+ * @return Property pointer
+ */
+std::unique_ptr<Property> loadProperty(::NeXus::File *file, const std::string &group,
```
Can we factor our portions of this member function that are the same as in `std::unique_ptr<Property> loadProperty(::NeXus::File *file, const std::string &group)`?

williamfgc committed:
Move to reference in non-owning functions

Replace std::shared_ptr as it is not needed.

williamfgc committed:
Refactor PropertyNexus::LoadProperty overloads

williamfgc committed:
Refactor Run::loadNexus to resuse common parts

Introduce Run::loadNexusCommon

peterfpeterson commented:
For the comment "reported speed ups on 20% on average on ORNL SNS/HFIR production system" can you give more details? e.g. before and after timing, and characteristics of files.

williamfgc committed:
Refactor MDBoxFlatTree::loadExperimentInfos

Clean unused variable in LoadMD

williamfgc commented:
@quantumsteve thanks for the review, I addressed your points. 
@peterfpeterson please see the attached image running on analysis.sns.gov :

<img width="781" alt="Screen Shot 2021-09-14 at 7 46 37 PM" src="https://user-images.githubusercontent.com/19393110/133838897-3fd9b7c7-34d6-4e12-97f3-d64a0a94f710.png">

williamfgc committed:
Add missing MDBoxBaseTest::loadAndAddFrom

williamfgc committed:
Address Doxygen checks failures

williamfgc commented:
It should be good to go.
